doc/reference/trd104-syscalls.md

```diff
@@ -509,19 +509,19 @@ kernel returns `Failure` with an error code of `NODEVICE`.
 ---------------------------------
 
 The Read-Write Allow system call class is how a userspace process
-shares a buffer with the kernel that the kernel can read and write. 
-
-Calling a Read-Write Allow system call returns a buffer (address and
-length).  On the first call to a Read-Write Allow system call, the
-kernel returns a zero-length buffer. Subsequent successful calls to
-Read-Write Allow return the previous buffer passed.  The standard
-access model for allowed buffers is that userspace does not read or
-write a buffer that has been allowed: access to the memory is intended
-to be exclusive either to userspace or to the kernel. To regain access
-to a passed buffer, the process calls the same Read-Write Allow system
-call again, which, if successful, returns the buffer. It can do so
-with a zero-length buffer if it wishes to pass no memory to the
-kernel. 
+shares a buffer with the kernel that the kernel can read and write.
+
+The standard access model for allowed buffers is that userspace does
+not read or write a buffer that has been allowed: access to the memory
+is intended to be exclusive either to userspace or to the kernel. To
+regain access to a passed buffer, the process calls the same
+Read-Write Allow system call again, which, if successful, returns the
+buffer. It can do so with a zero-length buffer if it wishes to pass no
+memory to the kernel. Once a buffer has been returned to userspace as
+part of a Read-Write Allow system call, it is guaranteed for the
+kernel to no longer have access to the described memory region, unless
+it is currently shared with the kernel as part of the passed in buffer
+or another Allow mechanism.
```

Comment:
The way I read this, it is allowed for a process to replace an Allowed buffer with a different buffer which overlaps the previous buffer. Was that intentional?

If we choose to implement an overlap check, it ends up being a little awkward. There are either two ways to implement the check:

1. Ignore the existing buffer with the same driver and allow number when checking for overlaps.
2. Remove the existing buffer from the overlap check data structure, check for overlap with the new buffer, and if the new buffer overlaps (i.e. the allow fails) re-add the existing buffer to the data structure.

Not that that's unreasonable, but it would add a bit of code size.

Reply:
Yes, that was intentional. I think that #2632 we committed to not introducing overlap checks?

Reply:
I thought we were allowing overlap checks to be added in the future. If we don't allow overlap checks, then this makes more sense.

Reply:
Well, now I'm not sure any longer myself :smile:. You might have a point. @phil-levis @hudson-ayers was that something we wanted to leave open for the future, or do we commit to just say "a process SHOULD NOT do this, but can technically"?

Reply:
The conclusion was that we should allow overlapped Allows at the system call level. If userspace wants to not allow then though its libraries, that's OK, but then it's a userspace check. After discussion, the conclusion was that the cycle and code size cost of the checks were not worth it given the API protections for memory safety.